rfc/rfc-0000-common-blueprint.md

```diff
@@ -0,0 +1,220 @@
+# Meta
+[meta]: #meta
+- Name: Common Blueprint Architecture
+- Start Date: 2022-03-29
+- Author(s): @emmjohnson @martyspiewak
+- Status: Draft
+- RFC Pull Request: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This RFC proposes collapsing all current "Blueprints" and templates (not including `ClusterRunTemplate`) into a common type, `ClusterBlueprint`. 
+As part of this, the RFC proposes that outputs be dynamically typed via a new CR `ClusterBlueprintType` and remove the statically typed outputs that exist today (url, revision, image, config). This allows us to maintain the contract of swappable templates, without restricting outputs to a predefined set. 
+Additionally, the selectors will be removed from blueprints and instead be specified in a new CR `ClusterBlueprintSelector`.
+
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+
+  - Today we have seven CRDs that are needed to create a supply chain and delivery. To simplify the data model for users, we can
+collapse all seven CRDs into a few common CRDs. Additionally, whereas today the overhead to start reconciling templates would be adding five identical
+reconcilers, this would enable us to have one reconciler for all blueprints and more easily add meaningful data to their statuses.
+
+- What use cases does it support?
+
+  - This would allow for snippets and users to specify any output they want, while maintaining strong typing.
+  - The `ClusterBlueprintSelector` will enable users to separate RBAC between template authors and operators who want to restrict what blueprints can be selected.
+
+- What is the expected outcome?
+
+  - Simplified on-boarding and user experience.
+
+# What it is
+[what-it-is]: #what-it-is
+* `ClusterBlueprints` will encapsulate everything the current templates allow, while adding `spec.resources` to enable defining multiple resources (i.e. snippets).
+* A `ClusterBlueprint` can take one of `spec.resources`, `spec.template`, and `spec.ytt`. It will allow inlining of templates in `spec.resources`.
+* A `ClusterBlueprint` must specify a `blueprintTypeRef` which defines the outputs required from the blueprint. The `ClusterBlueprint`'s `spec.outputs` must match those outputs.
```

Comment:
today we have those four CRDs

- Cluster**Source**Template, for exposing **source** info (`{url: <string>, revision: <string>}`[^1])
- Cluster**Image**Template, for exposing **image** info (`{image: <string>}`)
- Cluster**Config**Template, for exposing **config** info (`{config: <map[string]interface{}}}`)
- *ClusterTemplate, not exposing anything* 

So, I'd imagine that those first three would look like:

```yaml
---
apiVersion: carto.run/v1alpha1
kind: ClusterBlueprintType
metadata:
  name: source
spec:
  outputs:
    - name: url
    - name: revision

---
apiVersion: carto.run/v1alpha1
kind: ClusterBlueprintType
metadata:
  name: image
spec:
  outputs:
    - name: image

---
apiVersion: carto.run/v1alpha1
kind: ClusterBlueprintType
metadata:
  name: config
spec:
  outputs:
    - name: config
```

but then the fourth (`ClusterTemplate`), how would that look like? 

```yaml
apiVersion: carto.run/v1alpha1
kind: ClusterBlueprintType
metadata:
  name: no-outputs
spec:
  outputs: []
```

?

?

---
[^1]: _(in the comment above ^ I added some types for the fields, but, those are not really enforced as of today)_

Reply:
I think `ClusterTemplate` should be defined how you described so that it is explicit in the blueprint. I don't think we should change the behavior of a blueprint to  have to expose something.